routing/DELEGATED_NAMING_HTTP.md

```diff
@@ -0,0 +1,176 @@
+# Delegated Naming HTTP API
+
+![wip](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square)
+
+**Author(s)**:
+
+- [Masih H. Derkani](https://github.com/masih)
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+Delegated naming is a mechanism for IPFS implementations to offload naming system to another process
+or server. This includes naming resolution as well as publication of new naming records. This
+document describes an HTTP API through which such functionality is facilitated.
+
+## API Specification
+
+The Delegated Naming System HTTP API uses the `application/json` content type by default.
+
+As such, human-readable encodings of types are preferred. This spec may be updated in the future
+with a compact `application/cbor` encoding, in which case compact encodings of the various types
+would be used.
+
+## Common Data Types
+
+### IPNS Record
+
+The following snippet outlines the JSON schema of IPNS records:
+
+```json
+{
+  "Signature": "<signature>",
+  "Payload": {
+    "Value": "<value>",
+    "Sequence": 0,
+    "Validity": {
+      "EOL": {
+        "Timestamp": 0,
+        "AdvisoryTTL": 0
+      }
+    },
+    "PublicKey": "<optional-public-key>",
+    "ExtendedData": {}
+  }
+}
+```
+
+Where:
+
+- `Signature` is the multibase-encoded signature of the sha256 hash of the `Payload` field, signed
+  using the private key that corresponds to the `PublicKey` in the `Payload` if present. And
+  Otherwise, the private key associcated to the IPNS record key. Signing details for specific key
+  types should
+  follow [libp2p/peerid specs](https://github.com/libp2p/specs/blob/master/peer-ids/peer-ids.md#key-types),
+  unless stated otherwise.
+- `Payload` is the content of the IPNS record as specified
+  by [IPNS Record](../ipns/IPNS.md#ipns-record) specification:
+- `Value` is the string representation of the IPNS path,
+  e.g. `ipns/{ipns-key}`, `/ipns/example.com`, `/ipfs/baf...`, etc.
+- `Sequence` represents the current version of the record starting from `0`.
+- `Validity` captures the mechanism by which the record is validated. Each validity type reserves a
+  field key under this object.
+  - `EOL` donates that the validity type is EOL, containing:
+    - `Timestamp` represents the time in the future at which the record expires with nanoseconds
+       precision represented as an ASCII string that follows notation
+       from [RFC3339](https://www.ietf.org/rfc/rfc3339.txt).
+    - `AdvisoryTTL` represents an optional field that hints at how long the record should be
+       cached.
```

Comment:
> Someone will see resolved path and will use Value without checking signature.

* It is the delegated servers reponsibility to validate the records. Happy to spell that out if it's not obvious from the specification.
* The rationale is that the server should also offer enough information for the client to be able to verify records independently should the client choose to.

> We should remove these clear text fields, nudging people to use validation libraries that return a value only if the signature is valid.

This is certainly a way to do it but seems extreme at the cost of human readability, cross language extensibility. I need your help to understand why binary format would encourage people to call `.Validate` instead of decoding and using fields straight up?

Reply:
> why binary format would encourage people to call .Validate instead of decoding and using fields straight up?

Cleartext means the fastest way to make things work is to do the wrong thing – use Value from JSON response and don't care about signature validation.

Binary requires an extra step, which  means someone had to find go/js-ipns library, read its API/docs, and that increases the chance they will see `.Validate` function and at least THINK for a second why it is there.

But as I said, not feeling strongly, if IPNI feels JSON should maximize human readability, I can live with that, as long we have mandatory  `application/vns.ipfs.ipns-record` too.

> It is the delegated servers reponsibility to validate the records

My understanding is the only value delegated server validation provides is during PUTs: it protects IPNI from storing invalid records, and informs client about typos/errors.

During GET, clients can't blindly trust HTTP server returned truth to them, because they delegate verification to at least two entities (Protocol Labs AND the CA company that signed TLS cert for HTTPS). 

Using IPNS GET result from IPNI without verifying signature is as bad as not verifying CID against received block or using a random public HTTP gateway.